src/recorder.jl

```diff
@@ -29,111 +21,99 @@ https://dev.fast.ai/learner#Recorder
 =#
 
 """
-    Recorder
-    Recorder(learn::Learner; train_loss = true, train_smooth_loss = true,
-                             validate_loss = true, validate_smooth_loss = true)
-
-Container for [`Learner`](@ref) statistics (e.g. lr, loss and metrics) during training.
-Statistics are indexed by name, epoch and batch.
-For example to get the smoothed training loss for epoch 2, batch 3, we would call
-```julia
-recorder["TrainSmoothLoss", 2, 3]
-```
-To get the entire history of smooth training loss, one would call
-```julia
-recorder["TrainSmoothLoss", :, :]
-```
-"""
-struct Recorder
-    learn::Learner
-    logs::Dict{String,Array{Any,2}}
-end
-function Recorder(learn::Learner; train_loss = true, train_smooth_loss = true,
-                                  validate_loss = true, validate_smooth_loss = true)
-    rec = Recorder(learn, Dict{String,Array}())
-    if train_loss
-        add_cb!(learn,TrainLossRecorder(rec))
-    end
-    if train_smooth_loss
-        add_cb!(learn,TrainSmoothLossRecorder(rec))
-    end
-    if validate_loss
-        add_cb!(learn,ValidateLossRecorder(rec))
-    end
-    if validate_smooth_loss
-        add_cb!(learn,ValidateSmoothLossRecorder(rec))
-    end
-    return rec
-end
+    TrainLossRecorder
 
+Record a log of training loss
 """
-    add!(rec::Recorder, name, epoch_count, batch_size)
+struct TrainLossRecorder <: AbstractCallback
+    log::Array{Real,2}    
+end
 
-Add `name` to `rec` to be tracked.
-"""
-add!(rec::Recorder, name, epoch_count, batch_size) = rec.logs[name] = fill(nothing,epoch_count,batch_size)
+before_fit(lr::TrainLossRecorder,lrn::Learner, epoch_count, batch_size) = lr.log = zeros((epoch_count,batch_size))
+batch_train_loss(lr::TrainLossRecorder,lrn::AbstractLearner, epoch, batch, loss) = lr.log[epoch,batch] = loss
+Base.getindex(lr::TrainLossRecorder,idx...) = lr.log[idx]
 
 """
-    log!(rec::Recorder, name::String, epoch::Int, batch::Int, value)
+    ValidateLossRecorder
 
-Log `value` to `rec[name, epoch, batch]`.
+Record a log of validation loss
 """
-function log!(rec::Recorder, name::String, epoch::Int, batch::Int, value)
-#    println("$name,$epoch,$batch,$value")
-    rec.logs[name][epoch,batch] = value
+struct ValidateLossRecorder <: AbstractCallback
+    log::Array{Real,2}
 end
 
-Base.getindex(rec::Recorder,idx...) = rec.logs[idx[1]][idx[2],idx[3]]
-Base.setindex!(rec::Recorder,value,idx...) = rec.logs[idx[1]][idx[2],idx[3]] = value
+before_fit(lr::TrainLossRecorder,lrn::Learner, epoch_count, batch_size) = lr.log = zeros((epoch_count,batch_size))
```

Comment:
Looks like you have a copy-paste typo. Should be `lr::ValidateLossRecorder`.

Reply:
Thanks.  Fixed